core/src/main/java/com/datastax/oss/driver/internal/core/type/codec/registry/CachingCodecRegistry.java

```diff
@@ -273,15 +277,10 @@ private boolean collides(TypeCodec<?> newCodec, TypeCodec<?> oldCodec) {
       }
     }
 
-    if (value instanceof TupleValue) {
-      return uncheckedCast(codecFor(((TupleValue) value).getType(), TupleValue.class));
-    } else if (value instanceof UdtValue) {
-      return uncheckedCast(codecFor(((UdtValue) value).getType(), UdtValue.class));
-    }
-
-    GenericType<?> javaType = inspectType(value, null);
+    DataType cqlType = inferCqlTypeFromValue(value);
+    GenericType<?> javaType = inspectType(value, cqlType);
```

Comment:
It feels a bit awkward to have two separate methods here, since they both perform the same traversal of the value.

We could have a single method that constructs both CQL type and Java type at the same type, but it would force us to duplicate `inspectType`, and also we'd need a pair return type which is awkward in Java.

So all in all it's fine the way it is I guess.